### PR TITLE
Update generated lextab/parsetab files

### PR DESCRIPTION
**Pull request by astrofrog**
This does the following:

* Updates the bundled version of PLY to 3.10
* Updates the code that auto-generates the lexers/parsers to add a header line with the license (which used to be handled by ``clean_parse_tables.py``
* Updates the comments to explain how to regenerate the file, and include that in the header of the lexers/parsers
* Re-generates the files

Just to be clear, this does *not* use ``clean_parse_tables.py`` anymore.

Now one thing that isn't ideal is that in Python 3, it's more obvious that e.g. sets are returned in a non-deterministic order, so these files will change every time they are re-generated, which is not ideal. Some of it is under our control (see the addition of ``sorted`` for one of the sets), but I'm not sure how to fix them all. This may be something that has to be fixed in ply itself.

I think this should fix https://github.com/astropy/astropy/issues/7068

### CODE REVIEW

File: astropy/units/format/cds.py [hunk 1]
```diff
@@ -120,9 +126,15 @@ def t_error(t):
             raise ValueError(
                 "Invalid character at col {0}".format(t.lexpos))
 
+        lexer_exists = os.path.exists(os.path.join(os.path.dirname(__file__),
+                                      'cds_lextab.py'))
+
         lexer = lex.lex(optimize=True, lextab='cds_lextab',
                         outputdir=os.path.dirname(__file__),
-                        reflags=re.UNICODE)
+                        reflags=int(re.UNICODE))
```
Comment (astrofrog) [inline astropy/units/format/cds.py]: @eteq @bsipocz - technically speaking this is the actual thing that was causing the bug in https://github.com/astropy/astropy/issues/7068 - should I split this out into a separate PR to backport to v2.0.x? Or should I do a separate PR to v2.0.x with this minimal fix?

**Comment by astrofrog** [comment]
Just for info, I opened a PR to ply to make the generated files deterministic (i.e. they will be the same every time they are re-generated): https://github.com/dabeaz/ply/pull/151

File: astropy/units/format/cds.py (hunk 1 above)
Comment (bsipocz) [reply astropy/units/format/cds.py]: Ohh, good catch. 

I would say both approach is fine with me, but with the aim to get 3.0 out of the door, maybe the latter is better.

**Comment by bsipocz** [comment]
Updating ply to 3.10 actually preproduced the same error from #7068, so since this one passes I'm certain that that issue is solved by this PR. Thanks @astrofrog!

**Comment by astrofrog** [comment]
By the way the ``re`` flag issue was actually due to a bug in PLY in Python 3.6, which I've opened a PR for here: https://github.com/dabeaz/ply/pull/153

**Comment by astrofrog** [comment]
@eteq - does this satisfy https://github.com/astropy/astropy/issues/4318 or do you *also* want a section in the dev docs themselves? I feel like having the instructions inside the files themselves and at the top of the Python files might be enough because stuff just gets lost in the vast dev docs...

**Comment by bsipocz** [comment]
@astrofrog - I think, for this case, having the instructions in the files themselves are more useful than having it somewhere in the docs.

File: astropy/units/format/cds.py (hunk 1 above)
Comment (eteq) [reply astropy/units/format/cds.py]: I agree a separate PR with just the minimal fix.

File: astropy/coordinates/angle_utilities.py [hunk 2]
```diff
@@ -24,6 +29,20 @@
...
 
 
+TAB_HEADER = """# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# This file was automatically generated from ply. To re-generate this file,
+# remove it from this folder, then build astropy and run the tests in-place:
+#
+#   python setup.py build_ext --inplace
+#   pytest astropy/coordinates
```
Comment (eteq) [inline astropy/coordinates/angle_utilities.py]: how about replacing this test part with: 
```
python -c "from astropy.coordinates import angle_utilities;angle_utilities._AngleParser()"
```
?  That seems a bit more fine-grained

File: astropy/units/format/base.py [hunk 3]
```diff
@@ -18,6 +21,20 @@ def __new__(mcls, name, bases, members):
...
 
 
+TAB_HEADER = """# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# This file was automatically generated from ply. To re-generate this file,
+# remove it from this folder, then build astropy and run the tests in-place:
+#
+#   python setup.py build_ext --inplace
+#   pytest astropy/units
```
Comment (eteq) [inline astropy/units/format/base.py]: This one I don't see an obvious way to replace it with a more targeted call...

File: astropy/coordinates/angle_utilities.py (hunk 2 above)
Comment (astrofrog) [reply astropy/coordinates/angle_utilities.py]: Mine is more general in case we add multiple parsers in future :)

Comment (eteq) [reply astropy/coordinates/angle_utilities.py]: Hmm, I think it's very dangerous to use the tests to do this... but I see your point that *something* more general is needed.  So I guess this is fine for consistency, but I'm going to make an issue to try to address this better anyway.
